File: frontend/src/utils/frontmatter.ts
/**
 * Wiki page metadata extracted from frontmatter.
 * All fields are optional as not all pages may have complete metadata.
 */
export interface WikiPageMeta {
  /** Path to the source file/directory this page documents */
  source?: string
  /** Type of documentation page: 'file', 'directory', etc. */
  type?: string
  /** ISO timestamp when the page was generated */
  generated?: string
  /** Git commit hash at generation time */
  commit?: string
  /** Architectural layer (e.g., 'api', 'core', 'ui') */
  layer?: string
}

/** Known frontmatter fields to extract */
const KNOWN_FIELDS: ReadonlySet<string> = new Set([
  'source',
  'type',
  'generated',
  'commit',
  'layer',
])

/**
 * Parse YAML-style frontmatter from wiki page content.
 *
 * Frontmatter must:
 * - Start at the very beginning of the content
 * - Be delimited by `---` on its own line
 * - Contain simple `key: value` pairs (one per line)
 *
 * @param content - The full page content including potential frontmatter
 * @returns Object with extracted metadata (or null) and the content without frontmatter
 *
 * @example
 * ```ts
 * const { meta, content } = parseFrontmatter(`---
 * source: path/to/file.py
 * type: file
 * ---
 *
 * # Page Title`)
 *
 * // meta = { source: 'path/to/file.py', type: 'file' }
 * // content = '# Page Title'
 * ```
 */
export function parseFrontmatter(content: string): {
  meta: WikiPageMeta | null
  content: string
} {
  // Frontmatter must start at the beginning with ---
  if (!content.startsWith('---')) {
    return { meta: null, content }
  }

  // Find the closing ---
  // Look for \n---\n or \n--- at end of string
  const closingMatch = content.match(/\n---(\r?\n|$)/)
  if (!closingMatch || closingMatch.index === undefined) {
    return { meta: null, content }
  }

  const closingIndex = closingMatch.index
  const frontmatterBlock = content.slice(4, closingIndex) // Skip opening ---\n
  const afterFrontmatter = content.slice(closingIndex + closingMatch[0].length)

  // Parse key: value pairs from frontmatter block
  const meta: WikiPageMeta = {}
  const lines = frontmatterBlock.split('\n')

  for (const line of lines) {
    const trimmedLine = line.trim()
    if (!trimmedLine) continue

    // Find the first colon to split key: value
    const colonIndex = trimmedLine.indexOf(':')
    if (colonIndex === -1) continue

    const key = trimmedLine.slice(0, colonIndex).trim()
    const value = trimmedLine.slice(colonIndex + 1).trim()

    // Only include known fields
    if (KNOWN_FIELDS.has(key)) {
      meta[key as keyof WikiPageMeta] = value
    }
  }

  // Trim leading whitespace from content (but preserve content structure)
  const trimmedContent = afterFrontmatter.replace(/^[\r\n]+/, '')

  return { meta, content: trimmedContent }
}

File: backend/src/oya/generation/frontmatter.py
"""Utilities for building and parsing YAML frontmatter in wiki pages.

Frontmatter provides metadata for each wiki page including:
- source: The source file/directory path (optional for overview pages)
- type: Page type (file, directory, overview, etc.)
- commit: Git commit hash when page was generated
- generated: ISO timestamp of generation
- layer: Architecture layer (optional)

This metadata enables detection of orphaned pages during cleanup.
"""

from datetime import datetime

import yaml


def build_frontmatter(
    source: str | None,
    page_type: str,
    commit: str,
    generated: datetime,
    layer: str | None = None,
) -> str:
    """Build YAML frontmatter string for a wiki page.

    Args:
        source: Source file/directory path (None for overview pages)
        page_type: Type of page (file, directory, overview, etc.)
        commit: Git commit hash when page was generated
        generated: Datetime when page was generated
        layer: Architecture layer (optional)

    Returns:
        YAML frontmatter string starting with --- and ending with ---
        followed by a blank line.
    """
    metadata: dict[str, str] = {}

    if source is not None:
        metadata["source"] = source

    metadata["type"] = page_type
    metadata["generated"] = generated.isoformat()
    metadata["commit"] = commit

    if layer is not None:
        metadata["layer"] = layer

    # Build YAML with explicit string formatting to control order
    lines = ["---"]
    for key, value in metadata.items():
        lines.append(f"{key}: {value}")
    lines.append("---")
    lines.append("")  # Blank line after frontmatter

    return "\n".join(lines) + "\n"


def parse_frontmatter(content: str) -> tuple[dict | None, str]:
    """Parse YAML frontmatter from wiki page content.

    Args:
        content: Full page content that may start with frontmatter

    Returns:
        Tuple of (metadata_dict, remaining_content).
        If no valid frontmatter found, returns (None, original_content).
    """
    if not content.startswith("---\n"):
        return None, content

    # Find the closing delimiter
    # Start searching after the opening "---\n"
    end_pos = content.find("\n---\n", 4)
    if end_pos == -1:
        # Check for closing delimiter at end of content
        if content.rstrip().endswith("\n---"):
            end_pos = content.rstrip().rfind("\n---")
        else:
            return None, content

    # Extract the YAML content between delimiters
    yaml_content = content[4:end_pos]

    try:
        metadata = yaml.safe_load(yaml_content)
        if not isinstance(metadata, dict):
            return None, content
    except yaml.YAMLError:
        return None, content

    # Calculate where remaining content starts
    # After closing "---\n", skip optional blank line
    remaining_start = end_pos + 5  # len("\n---\n")

    # Skip one blank line if present
    if remaining_start < len(content) and content[remaining_start] == "\n":
        remaining_start += 1

    remaining_content = content[remaining_start:]

    return metadata, remaining_content
